Listbox and tree: clarify requirements for selected and checked

mcking65 opened the pull request:
This PR will resolve #700 and resolve #1052. It supercedes #799.

Changes include the following.

Removes "selectable" from the definition of option to support cases where either an item is checkable instead of selectable or an item is neither selectable nor checkable (i.e., similar to disabled). The definition of option is then:

>An item in a listbox.

Similarly simplifies the definition of treeitem.

Replaces the statement:

>Elements with the role option have an implicit aria-selected value of false.

with a list of conditions that must be met before a user agent can provide an implicit value for aria-selected.

Adds the same restrictions on user agent provision of implicit aria-selected values to treeitem and tab.

These user agent restrictions enable authors to:
1. Use aria-checked instead of aria-selected in a listbox or tree.
2. Indicate that an option or treeitem is not selectable or checkable by setting checked or selected to undefined.

Adds authoring guidance that:
1. Weakly suggests use of checked instead of selected in multi-select listboxes and trees.
2. Strongly discourages use of both checked and selected in the same listbox or tree.

In the characteristics table for option:
1. Removes aria-selected from required properties and adds to supported properties.
2. Removes the implicit value for aria-selected.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):
#option, #treeitem, #tab, #aria-selected
    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Jan 20, 2021, 10:59 PM UTC)_.

<details>
<summary>More</summary>


PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [Spec Generator](https://www.w3.org/2015/labs/) - Spec Generator is the web service used to build specs that rely on ReSpec.

:link: [Related URL](https://labs.w3.org/spec-generator/?type=respec&url=https%3A%2F%2Frawcdn.githack.com%2Fw3c%2Faria%2Fc51f1486be17adf29d164be548147d860232aa5f%2Findex.html%3FisPreview%3Dtrue)

```
Navigation timeout of 26205 ms exceeded
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20w3c/aria%231340.)._
</details>

carmacleod commented on index.html:
```diff
@@ -5991,9 +5991,26 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
+        </p>
+        <ul>
+          <li>The meaning and purpose of <sref>aria-selected</sref> is different from the meaning and purpose of <sref>aria-checked</sref> in the user interface.</li>
+          <li>The user interface makes the mening and purpose of each state apparent.</li>
+          <li>The user interface provides a separate method for controling each state.</li>
```
Couple of typos:
```suggestion
          <li>The user interface makes the meaning and purpose of each state apparent.</li>
          <li>The user interface provides a separate method for controlling each state.</li>
```

carmacleod commented:
I like this! 🎉 
Thanks @mcking65, @aleventhal and @smhigley !

carmacleod commented:
Still need to fix up the wording in [aria-selected](https://pr-preview.s3.amazonaws.com/w3c/aria/1340/02fee23...2929945.html#aria-selected) to match these new words.

This sentence is particularly problematic:
> The selection normally follows the focus, and is managed by the user agent.

It should either say "may be managed by the user agent", or be completely deleted.

carmacleod commented:
(Accidentally clicked the wrong button and closed this - have reopened.)

mcking65 commented:
This is now ready for working group review. Please see the top comment for a complete summary of changes.

jnurthen commented:
@jcsteh can you please review this. Thanks.

cookiecrook commented:
Matt King requests reviewers re-read conversation in #1052 before providing PR review.

carmacleod commented:
> The option role supports both aria-selected and aria-checked because it is recommended that authors indicate selection with aria-selected in single-select list boxes and with aria-checked in multi-select list boxes.

This is no longer "Weakly suggesting use of checked instead of selected in multi-select listboxes and trees."

I preferred the previous wording:
> ... because it is common for authors to use selection in single-select list boxes and check marks in multi-select list boxes.

Are we certain that we want to make this a strong recommendation?

smhigley commented on index.html:
```diff
@@ -5991,9 +5991,26 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
+        </p>
+        <p>
+          The <rref>option</rref> role supports both <sref>aria-selected</sref> and <sref>aria-checked</sref> because it is recommended that authors indicate selection with <sref>aria-selected</sref> in single-select list boxes and with <sref>aria-checked</sref> in multi-select list boxes.
+          Authors SHOULD NOT specify both <sref>aria-selected</sref> and <sref>aria-checked</sref> on <rref>option</rref> elements contained by the same <rref>listbox</rref> except in the extremely rare circumstances where all the following conditions are met:
```
My one suggestion would be to tweak the "except when" wording, since (colloquially) it currently implies that authors *should* use both selected and checked when they think those three conditions are met.

Could we split up this line into two sentences, something like "Authors SHOULD NOT specify both aria-selected and aria-checked on option elements contained by the same listbox. However, authors MAY use both in the extremely rare circumstances where all the following conditions are met: [...]" (though probably with some more wording tweaks)

mcking65 replied:
Fixed with 03732dc

smhigley replied:
@mcking65 I like the change in that commit a lot! I think that commit might've gotten lost along the way, though -- I don't see it in this PR right now

smhigley reviewed (approved):
I just left one comment, but I think overall this will be a really valuable update 🎉

jnurthen commented:
@aleventhal @jcsteh @cookiecrook please review as would love to merge this

carmacleod reviewed (changes requested):
> The treeitem role supports both aria-selected and aria-checked because it is recommended that authors indicate selection with aria-selected in single-select trees and with aria-checked in multi-select trees.

Note that the new wording for `treeitem` (and for `option`, as mentioned in https://github.com/w3c/aria/pull/1340#issuecomment-722538496) uses the word "recommended", which is an [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt) word equivalent to **SHOULD**.

I think we need to be careful about saying that authors SHOULD use aria-checked instead of aria-selected in multi-select trees and listboxes. Many (most?) of the multi-select trees and listboxes I've seen do not use checkboxes on the items. For example:
- File Explorer's listview (default does not have checkboxes),
- Finder's "list" view is a tree,
- VSCode file tree, and other IDE file trees (desktop and online).

The only time I can think of where checkboxes make more sense than selection in a multi-select listbox or tree is when the control represents a bunch of user settings, for example:
- macOS Settings - Keyboard (Services is a checkbox tree, other keyboard categories are checkbox lists),
- Windows Control Panel "Windows Features" dialog has a checkbox tree.

jcsteh commented on index.html:
```diff
@@ -9891,9 +9908,27 @@ <h5>Presentational Roles Conflict Resolution</h5>
-				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>treeitem</code> are contained in, or <a>owned</a> by, an element with the role <rref>group</rref> or <rref>tree</rref>.</p>
+				<p>An item in a <rref>tree</rref>.</p>
+        				<p>A <rref>treeitem</rref> <a>element</a> may contain a sub-level group of elements that can be expanded or collapsed. An expandable collection of <code>treeitem</code> elements are enclosed in an element with the <rref>group</rref> <a>role</a>.</p>
+				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>treeitem</code> are contained in, or <a>owned</a> by, an element with role <rref>tree</rref> or an element with role <rref>group</rref> that is contained in, or owned by, and element with role <rref>treeitem</rref>.</p>
```
typo
```suggestion
				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>treeitem</code> are contained in, or <a>owned</a> by, an element with role <rref>tree</rref> or an element with role <rref>group</rref> that is contained in, or owned by, an element with role <rref>treeitem</rref>.</p>
```

jcsteh commented on index.html:
```diff
@@ -8694,11 +8708,14 @@ <h5>Presentational Roles Conflict Resolution</h5>
+        <p>User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>tab</rref> in a <rref>tablist</rref> if the following conditions are met, but user agents MUST NOT provide an implicit value for <sref>aria-selected</sref> if any of the following conditions are not met:</p>
+        <ul>
+          <li>The value of <pref>aria-multiselectable</pref> on the <rref>tablist</rref> is <code>false</code> or <code>undefined</code>.</li>
+          <li>None of the <rref>tab</rref> elements in the <rref>tablist</rref> have an explicitly declared value for <sref>aria-selected</sref> or <sref>aria-expanded</sref>.</li>  
```
It's probably silly to use aria-expanded in a single-select tablist. However, I think this sentence means that if aria-expanded is set in a single-select tablist, there is no implicit aria-selected. Why does aria-expanded have this effect in this case?

mcking65 replied:
Same as the argument for aria-checked similarly effecting a listbox.

If the author is using aria-expanded, we assume it is in lieu of aria-selected. If it is not, then it is up to the author to explicitly declare both. We don't want the user agent to assume the author wants both.

jcsteh reviewed (commented):
I agree with @carmacleod's [concerns about saying authors should use aria-checked instead of aria-selected in multi-select listboxes/trees](https://github.com/w3c/aria/pull/1340#pullrequestreview-533918960). This might be what we want to happen in future (and there are even some conversations about changing this in the HTML select widget), but as far as I've seen, this isn't what's happening today.

cookiecrook commented on index.html:
```diff
@@ -5991,9 +5991,26 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
-				<p>Elements with the role <code>option</code> have an implicit <sref>aria-selected</sref> value of <code>false</code>.</p>
+				<p>An item in a <rref>listbox</rref>.</p>
+				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>option</code> are contained in, or <a>owned</a> by, an element with the <a>role</a> <rref>listbox</rref> or <rref>group</rref> within a <code>listbox</code>. Options not associated with a <rref>listbox</rref> might not be correctly mapped to an <a>accessibility <abbr title="Application Programing Interfaces">API</abbr></a>.</p>
+        <p>User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>option</rref> in a <rref>listbox</rref> if the following conditions are met, but user agents MUST NOT provide an implicit value for <sref>aria-selected</sref> if any of the following conditions are not met:</p>
```
Editorial nit. Is is possible to rephrase some of the negative inversions: "…but user agents MUST **NOT** provide an implicit value [if conditions] are **not** met"… Also, some of the referenced list items start with "None of" so I'm having to triple invert when reading this. (A "None of" condition is "not met" so therefore the UA "must not" do what?)

Perhaps "MUST NOT provide an implicit value [unless one/some/all of the following conditions are true]" and then include related inversions of the bullets? (e.g. "One or more of" instead of "None of") Okay to keep the extra nots if inverting the list items becomes too problematic. I think it's accurate, just a little hard to read.

aleventhal replied:
How about "if and only if" the following conditions are met -- that seems clearer than what's there now.

mcking65 replied:
Fixed with 2df964f.
@aleventhal  I kept both the may only if and the then changed must not to must. I think it's important to have a MUST so the requirement is tested. Hopefully the new wording is much easier to read.

cookiecrook commented on index.html:
```diff
@@ -8694,11 +8708,14 @@ <h5>Presentational Roles Conflict Resolution</h5>
-
-				<p>In either case, authors SHOULD ensure that a selected tab has its <sref>aria-selected</sref> attribute set to <code>true</code>, that inactive tab elements have their <sref>aria-selected</sref> attribute set to <code>false</code>, and that the currently selected tab provides a visual indication that it is selected. In the absence of an <sref>aria-selected</sref> attribute on the current ta ...
+				<p>Authors SHOULD ensure that a selected tab has its <sref>aria-selected</sref> attribute set to <code>true</code>, that inactive tab elements have their <sref>aria-selected</sref> attribute set to <code>false</code>, and that the currently selected tab provides a visual indication that it is selected.</p>
+        <p>User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>tab</rref> in a <rref>tablist</rref> if the following conditions are met, but user agents MUST NOT provide an implicit value for <sref>aria-selected</sref> if any of the following conditions are not met:</p>
```
ditto re: "MUST NOT" if "none of" are "not met"

mcking65 replied:
Resolved by d97ef54

cookiecrook commented on index.html:
```diff
@@ -9891,9 +9908,27 @@ <h5>Presentational Roles Conflict Resolution</h5>
+				<p>An item in a <rref>tree</rref>.</p>
+        				<p>A <rref>treeitem</rref> <a>element</a> may contain a sub-level group of elements that can be expanded or collapsed. An expandable collection of <code>treeitem</code> elements are enclosed in an element with the <rref>group</rref> <a>role</a>.</p>
+				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>treeitem</code> are contained in, or <a>owned</a> by, an element with role <rref>tree</rref> or an element with role <rref>group</rref> that is contained in, or owned by, an element with role <rref>treeitem</rref>.</p>
+        <p>User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>treeitem</rref> in a <rref>tree</rref> if the following conditions are met, but user agents MUST NOT provide an implicit value for <sref>aria-selected</sref> if any of the following conditions are not met:</p>
```
ditto

mcking65 replied:
Resolved by d97ef54

aleventhal commented on index.html:
```diff
@@ -5991,9 +5991,26 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
+				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>option</code> are contained in, or <a>owned</a> by, an element with the <a>role</a> <rref>listbox</rref> or <rref>group</rref> within a <code>listbox</code>. Options not associated with a <rref>listbox</rref> might not be correctly mapped to an <a>accessibility <abbr title="Application Programing Interfaces">API</abbr></a>.</p>
+        <p>User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>option</rref> in a <rref>listbox</rref> if the following conditions are met, but user agents MUST NOT provide an implicit value for <sref>aria-selected</sref> if any of the following conditions are not met:</p>
+        <ul>
+          <li>The value of <pref>aria-multiselectable</pref> on the <rref>listbox</rref> is <code>false</code> or <code>undefined</code>.</li>
```
So this includes when the author literally has the attribute aria-multiselectable="undefined"

mcking65 replied:
Is that a problem? Should we word it so it does not? That would be some real hair splitting.

aleventhal commented on index.html:
```diff
@@ -5991,9 +5991,26 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
+        <p>User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>option</rref> in a <rref>listbox</rref> if the following conditions are met, but user agents MUST NOT provide an implicit value for <sref>aria-selected</sref> if any of the following conditions are not met:</p>
+        <ul>
+          <li>The value of <pref>aria-multiselectable</pref> on the <rref>listbox</rref> is <code>false</code> or <code>undefined</code>.</li>
+          <li>None of the <rref>option</rref> elements in the <rref>listbox</rref> have an explicitly declared value for <sref>aria-selected</sref> or <sref>aria-checked</sref>.</li>  
```
If there are 300 options, do you want us to go through each one and check? Or can the UA optimize a bit here?

mcking65 replied:
Good question.

I think we've figured out that if UA never provides implicit values, nothing will break in a way that harms users. So, one approach could be to limit conditions where implicit values are provided if you want a UA to provide them at all.

When building the AX tree, isn't it necessary to process every option and at the same time set its states and props? So, you might be doing this anyway just to create the tree.

aleventhal replied:
The main problem occurs when there are updates. We try to only process changed nodes, because as a page changes due to interactions, users expect things to be responsive. However, if we optimize so that once an implicit value is used, that it can be used going forward, then we should be ok on the browser performance front. I don't think the optimization needs any verbiage here, but we should clarify that it's a good approach.

aleventhal commented on index.html:
```diff
@@ -9891,9 +9908,27 @@ <h5>Presentational Roles Conflict Resolution</h5>
+        <p>User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>treeitem</rref> in a <rref>tree</rref> if the following conditions are met, but user agents MUST NOT provide an implicit value for <sref>aria-selected</sref> if any of the following conditions are not met:</p>
+        <ul>
+          <li>The value of <pref>aria-multiselectable</pref> on the <rref>tree</rref> is <code>false</code> or <code>undefined</code>.</li>
+          <li>None of the <rref>treeitem</rref> elements in the <rref>tree</rref> have an explicitly declared value for <sref>aria-selected</sref> or <sref>aria-checked</sref>.</li>  
```
Since is is basically the same text as for option, does it make sense to create a separate section for Multiselectable Container Widgets Supporting Checked or Selected, and then redirect the reader to there? That way we don't have to wonder if it's exactly the same.

aleventhal replied:
Also, that will make it easier to provide the same functionality in the future, e.g. if we ever add a listgrid role.

Finally, why are we only doing this for tree and listbox? What about treegrid? Can't rows be checked or selected there? Imagine a threaded email interface.

mcking65 replied:
I don't think it would simplify the wording to consolidate; it would still have to explicitly mention the option, listbox, tree, and treeitem roles.

In grid and treegrid, aria-checked is not supported on rows. In those widgets, if you were crazy enough to use both checked and selected states, you can just put  a checkbox in a cell. That is not possible in listbox and tree; you can't put a checkbox inside an option or treeitem.

carmacleod commented on index.html:
```diff
@@ -5991,9 +5991,28 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
-				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>option</code> are contained in, or <a>owned</a> by, an element with the <a>role</a>  <rref>listbox</rref> or <rref>group</rref> within a <code>listbox</code>. Options not associated with a <rref>listbox</rref> might not be correctly mapped to an <a>accessibility <abbr title="Application Programing Interfaces">API</abbr></a>.</p>
-				<p>Elements with the role <code>option</code> have an implicit <sref>aria-selected</sref> value of <code>false</code>.</p>
+				<p>An item in a <rref>listbox</rref>.</p>
+				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>option</code> are contained in, or <a>owned</a> by, an element with the <a>role</a> <rref>listbox</rref> or <rref>group</rref> within a <code>listbox</code>. Options not associated with a <rref>listbox</rref> might not be correctly mapped to an <a>accessibility <abbr title="Application Programing Interfaces">API</abbr></a>.</p>
```
For consistency with the nicely-updated clarifying words in treeitem, please change this sentence about groups within listboxes:

```suggestion
				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>option</code> are contained in, or <a>owned</a> by, an element with <a>role</a> <rref>listbox</rref> or an element with <a>role</a> <rref>group</rref> that is contained in, or <a>owned</a> by, an element with <a>role</a> <code>listbox</code>. Options not associated with a <rref>listbox</rref> might not be correctly mapped to an <a>accessibility <abbr title="Application Programing Interfaces">API</abbr></a>.</p>
```

mcking65 commented:
@carmacleod @jcsteh 

I hope the following language addresses your concerns about the language about the suitability of selected for single-selects and checked for multi-selects.

>Authors MAY indicate selection for option elements using either aria-selected or aria-checked. Some user interfaces indicate selection with aria-selected in single-select list boxes and with aria-checked in multi-select list boxes. Authors SHOULD NOT specify both aria-selected and aria-checked on option elements contained by the same listbox. However, authors MAY use both in the same listbox in the extremely rare circumstances where all the following conditions are met:
>
>* The meaning and purpose of aria-selected  is different from the meaning and purpose of aria-checked  in the user interface. 
>* The user interface makes the meaning and purpose of each state apparent. 
>* The user interface provides a separate method for controlling each state. 

Similar language is also in the definition of treeitem.

mcking65 commented:
@jnurthen @smhigley @carmacleod @aleventhal @jcsteh @cookiecrook 

I'm pretty sure I've addressed all feedback now. Please let me know if I've overlooked something or the changes are not satisfactory.

sinabahram commented:
@mcking65  thanks for the work on this.

We have this sentence:
Authors SHOULD NOT specify both aria-selected and aria-checked on option elements contained by the same listbox.

I do not agree with that sentence on its own in the case of multiselect. I thought our group consensus centered around limiting conveyance of selection, but not restricting both being used, since that's why there are criteria listed after that restriction for when both can be used.

Also, as written, it is a touch vague about whether we mean on the same option element or across option elements, plural, in the listbox.

Can we qualify that sentence to make it far less absolute and to only apply to the conveyance of selection, but not to restrict both being used on the same option element , as that was the point of the restriction carveouts?

For example:
Authors SHOULD NOT use both aria-selected and aria-checked to convey the same state on option elements contained by the same listbox.

That way, it is ok to use checked and selected for conveying different states e.g. the to-do list example with them being marked as done VS selected and marked as done as a group.

mcking65 commented:
@sinabahram commented:

> We have this sentence:
> Authors SHOULD NOT specify both aria-selected and aria-checked on option elements contained by the same listbox.
> 
> I do not agree with that sentence on its own in the case of multiselect. I thought our group consensus centered around limiting conveyance of selection, but not restricting both being used, since that's why there are criteria listed after that restriction for when both can be used.

That sentence does not stand on its own. It is a should not, not a must not. The following sentence contains a normative may statement and provides for the exception you want. The two sentences together read:

>Authors SHOULD NOT specify both aria-selected and aria-checked on option elements contained by the same listbox. However, authors MAY use both in the same listbox in the extremely rare circumstances where all the following conditions are met:

@sinabahram commented:

> Also, as written, it is a touch vague about whether we mean on the same option element or across option elements, plural, in the listbox.

You can use either or both properties on some or all of the options. I don't think it is necessary to explicitly spell that out. In other words, if the language doesn't explicitly prohibit something, it allows it.

carmacleod commented on index.html:
```diff
@@ -5991,9 +5991,31 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
+				<p>An item in a <rref>listbox</rref>.</p>
+				<p>Authors MUST ensure <a>elements</a> with <a>role</a> <code>option</code> are contained in, or <a>owned</a> by, an element with <a>role</a> <rref>listbox</rref> or an element with <a>role</a> <rref>group</rref> that is contained in, or <a>owned</a> by, an element with <a>role</a> <code>listbox</code>. Options not associated with a <rref>listbox</rref> might not be correctly mapped to an <a> ...
+        <p>
+          User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>option</rref> in a <rref>listbox</rref> only if the following conditions are met.
+          If a user agent provides an implicit value for <sref>aria-selected</sref> on an <rref>option</rref>, the user agent MUST ensure the following conditions are met before providing the implicit value:
```
It feels like there are too many words and sentences here, and some of them feel redundant.

Maybe reducing into one sentence and using "if and only if" (as suggested by @aleventhal earlier) can still have the same meaning?
```suggestion
          User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>option</rref> in a <rref>listbox</rref> if and only if the following conditions are met:
```

Alternatively, if you feel strongly that a MUST needs to follow the MAY, then perhaps the following would work?
```suggestion
          User agents MAY provide an implicit value for <sref>aria-selected</sref> for each <rref>option</rref> in a <rref>listbox</rref>, however the user agent MUST ensure the following conditions are met before providing the implicit value:
```

If you are ok with either of these suggestions, then the same fix would need to be made for treeitem and tab.

carmacleod commented on index.html:
```diff
@@ -5991,9 +5991,31 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
+          <li>None of the <rref>option</rref> elements in the <rref>listbox</rref> have an explicitly declared value for <sref>aria-selected</sref> or <sref>aria-checked</sref>.</li>  
+        </ul>
+        <p>
+          If a user agent provides an implicit <sref>aria-selected</sref> value for an <rref>option</rref>, the value SHOULD be <code>true</code> if the <rref>option</rref> has DOM focus or the <ref>listbox</ref> has DOM focus and the option is referenced by <pref>aria-activedescendant</pref>.
```
Need rref for listbox (not just ref). Also add rref for last instance of option:
```suggestion
          If a user agent provides an implicit <sref>aria-selected</sref> value for an <rref>option</rref>, the value SHOULD be <code>true</code> if the <rref>option</rref> has DOM focus or the <rref>listbox</rref> has DOM focus and the <rref>option</rref> is referenced by <pref>aria-activedescendant</pref>.
```

carmacleod commented on index.html:
```diff
@@ -9891,9 +9916,32 @@ <h5>Presentational Roles Conflict Resolution</h5>
+          <li>None of the <rref>treeitem</rref> elements in the <rref>tree</rref> have an explicitly declared value for <sref>aria-selected</sref> or <sref>aria-checked</sref>.</li>  
+        </ul>
+        <p>
+          If a user agent provides an implicit <sref>aria-selected</sref> value for a <rref>treeitem</rref>, the value SHOULD be <code>true</code> if the <rref>treeitem</rref> has DOM focus or the <ref>tree</ref> has DOM focus and the <rref>treeitem</rref> is referenced by <pref>aria-activedescendant</pref>.
```
Need rref for tree (not just ref):
```suggestion
          If a user agent provides an implicit <sref>aria-selected</sref> value for a <rref>treeitem</rref>, the value SHOULD be <code>true</code> if the <rref>treeitem</rref> has DOM focus or the <rref>tree</rref> has DOM focus and the <rref>treeitem</rref> is referenced by <pref>aria-activedescendant</pref>.
```

carmacleod reviewed (approved):
This is really close!
I only have a couple of suggestions left. :)
Giving an approving review anyhow, although I do feel that that 2-sentence MAY-MUST structure (in option, treeitem, and tab) could use some reduction for clarity.

mcking65 commented:
@sinabahram @carmacleod @smhigley 

Editorial changes based on your feedback are all complete now. Thank you for all the great feedback. All the wording seems much easier to understand now.

@joanmarie Note for testing: This PR introduces a new user agent must in option, treeitem, and tab. However, it is only a must for a UA that chooses to provide implicit values for aria-selected.

jnurthen commented:
@sinabahram @carmacleod @smhigley @cookiecrook @jcsteh @aleventhal this is ready to merge. Any objections please speak up before tomorrow's ARIA meeting.

aleventhal reviewed (approved):
I'm ok with this, but @mcking65 and @jcsteh, I'd like your feedback about by optimization comment. Namely that I want to avoid constantly checking whether there are any descendants that have aria-selected or aria-checked. I'd much prefer that if there ever are, we cache that info and never recompute it, and therefore the implicit selected decision would be based on a prior state.
